File: go/pkg/trees/trees_9.go
package trees

import (
	"fmt"
	"math/rand"
	"time"
)

// BSTNode is a node in a binary search tree
type BSTNode struct {
	Data      int
	Left      *BSTNode
	Right     *BSTNode
	LeftSize  int
	RightSize int
}

// ToString produces a string representation of the tree at n
func (n *BSTNode) ToString() string {
	if n == nil {
		return "nil"
	}
	return fmt.Sprintf("%d | %d | %d => (%s, %s)", n.Data, n.LeftSize, n.RightSize, n.Left.ToString(), n.Right.ToString())
}

// BST is a wrapper around the root node
type BST struct {
	Root *BSTNode
}

// ToString produces a string representation of the tree
func (t *BST) ToString() string {
	return t.Root.ToString()
}

// Insert a value into the tree
func (t *BST) Insert(data int) {
	if t.Root == nil {
		t.Root = &BSTNode{Data: data}
		return
	}

	node := t.Root
	var parent *BSTNode
	var direction string

	for node != nil {
		parent = node

		if data <= node.Data {
			parent.LeftSize++
			node = node.Left
			direction = "left"
		} else {
			parent.RightSize++
			node = node.Right
			direction = "right"
		}
	}

	if direction == "left" {
		parent.Left = &BSTNode{Data: data}
	} else {
		parent.Right = &BSTNode{Data: data}
	}
}

// Find a value in the tree
// return the found node or nil, the parent or nil, and the direction
// from the parent to the found node
func (t *BST) Find(data, decrement int) (*BSTNode, *BSTNode, string) {
	// empty tree
	if t.Root == nil {
		return nil, nil, ""
	}

	// root value is the value we're looking for
	if t.Root.Data == data {
		return t.Root, nil, ""
	}

	node := t.Root
	var parent *BSTNode
	var direction string

	leftDecrementedNodes := []*BSTNode{}
	rightDecrementedNodes := []*BSTNode{}

	for node != nil && node.Data != data {
		parent = node

		if data <= node.Data {
			parent.LeftSize = parent.LeftSize - decrement
			leftDecrementedNodes = append(leftDecrementedNodes, parent)
			node = node.Left
			direction = "left"
		} else {
			parent.RightSize = parent.RightSize - decrement
			rightDecrementedNodes = append(rightDecrementedNodes, parent)
			node = node.Right
			direction = "right"
		}
	}

	if node == nil && decrement > 0 {
		for _, decrementedNode := range leftDecrementedNodes {
			decrementedNode.LeftSize++
		}
		for _, decrementedNode := range rightDecrementedNodes {
			decrementedNode.RightSize++
		}
	}

	return node, parent, direction
}

// Delete a value from the tree
func (t *BST) Delete(data int) {
	node, parent, direction := t.Find(data, 1)

	if node == nil {
		return
	}

	var nodeToShift *BSTNode

	if node.Left != nil && node.Right != nil {
		nodeToShift = node.Left
		var parentOfNodeToShift *BSTNode

		for nodeToShift.Right != nil {
			parentOfNodeToShift = nodeToShift
			parentOfNodeToShift.RightSize--
			nodeToShift = nodeToShift.Right
		}

		// nodeToShift can only have a left subtree after the loop
		parentOfNodeToShift.Right = nodeToShift.Left
		nodeToShift.LeftSize = node.LeftSize - 1
		nodeToShift.RightSize = node.RightSize
	} else if node.Left != nil {
		nodeToShift = node.Left
		nodeToShift.LeftSize = node.LeftSize - 1
	} else if node.Right != nil {
		nodeToShift = node.Right
		nodeToShift.RightSize = node.RightSize - 1
	}
	// unwritten else case: leaf node (Left == nil && Right == nil)
	// safe to simply remove the node in this case

	// fix the size properties and ptrs on the parent
	if parent != nil {
		if direction == "left" {
			parent.Left = nodeToShift
		} else {
			parent.Right = nodeToShift
		}
	} else {
		// handle special case where the root is deleted
		t.Root = nodeToShift
	}

	if nodeToShift != nil {
		// set the children of the shifted node
		// to the deleted node it replaced
		// unless they are equivalent and would cause
		if node.Left != nodeToShift {
			nodeToShift.Left = node.Left
		}
		if node.Right != nodeToShift {
			nodeToShift.Right = node.Right
		}
	}
}

// Random returns a node uniformly at random from the tree
func (t *BST) Random() *BSTNode {
	if t.Root == nil {
		return nil
	}

	node := t.Root
	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	for node != nil {
		n := r.Float64()
		total := float64(1 + node.LeftSize + node.RightSize)

		if n <= 1/total {
			return node
		} else if n <= float64(1+node.LeftSize)/total {
			node = node.Left
		} else {
			node = node.Right
		}
	}

	// should be unreachable
	// could optionally panic here
	return nil
}

File: python/trees/trees_9.py
from random import random

# Approach:
# assuming that the data access patterns
# are such that any of the methods are as
# likely to be called as any other we should
# optimize for quick reads and writes
# one way to do this is to create a binary search tree

# for the implementation of random
# as long as each node knows how many nodes
# are in the left and right subtrees
# we can branch the probability of selection
# between left, right and stay accordingly


class BSTNode:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.right = None
        self.left_cardinality = 0
        self.right_cardinality = 0

    def __repr__(self):
        return f'{self.data} | {self.left_cardinality} | {self.right_cardinality} => ({self.left}, {self.right})'


class BST:
    def __init__(self, root=None):
        self.root = root

    def __repr__(self):
        return str(self.root)

    # O(log n) generally -- O(n) worst case
    def insert(self, data):
        if not self.root:
            self.root = BSTNode(data)
            return

        node = self.root

        while node:
            parent = node

            if data <= node.data:
                parent.left_cardinality += 1
                node = node.left
                direction = 'left'
            else:
                parent.right_cardinality += 1
                node = node.right
                direction = 'right'

        if direction == 'left':
            parent.left = BSTNode(data)
        else:
            parent.right = BSTNode(data)

    # O(log n) generally -- O(n) worst case
    # finds first node matching data value
    # provides an argument allowing delete to decrement
    def find(self, data, decrement=0):
        node = self.root

        if data == node.data:
            return node, None, None

        left_decremented_nodes = []
        right_decremented_nodes = []

        while node and node.data != data:
            parent = node
            if data <= node.data:
                parent.left_cardinality -= decrement
                left_decremented_nodes.append(parent)
                node = node.left
                direction = 'left'
            else:
                parent.right_cardinality -= decrement
                right_decremented_nodes.append(parent)
                node = node.right
                direction = 'right'

        # if a node to delete is not found
        # need to fix cardinalities
        if not node and decrement > 0:
            for decremented_node in left_decremented_nodes:
                decremented_node.left_cardinality += 1
            for decremented_node in right_decremented_nodes:
                decremented_node.right_cardinality += 1

        # will return None if data value does not exist in tree
        return node, parent, direction

    # O(log n) generally -- O(n) worst case
    def delete(self, data):
        node, parent, direction = self.find(data, decrement=1)
        node_to_shift = None

        # return None if data value is not found
        if not node:
            return

        if node.left and node.right:
            # find right most node of the left subtree
            node_to_shift = node.left

            while node_to_shift.right:
                parent_of_node_to_shift = node_to_shift
                parent_of_node_to_shift.right_cardinality -= 1
                node_to_shift = node_to_shift.right

            # place the left subtree of the shifting node in this node's place
            parent_of_node_to_shift.right = node_to_shift.left
            node_to_shift.left_cardinality = node.left_cardinality - 1
            node_to_shift.right_cardinality = node.right_cardinality
        elif node.left:

            # shift subtree up and to the right
            # in this case the parent is the node being deleted
            node_to_shift = node.left
            node_to_shift.left_cardinality = node.left_cardinality - 1
        elif node.right:
            # shift subtree up and to the left
            # in this case the parent is the node being deleted
            node_to_shift = node.right
            node_to_shift.right_cardinality = node.right_cardinality - 1

        # link parent to shifted subtree depending on direction found
        if parent:
            if direction == 'left':
                parent.left = node_to_shift
            else:
                parent.right = node_to_shift
        else:
            self.root = node_to_shift

        if node_to_shift:
            # link node_to_shift to original node's children
            # provided they are not the same
            if node.left is not node_to_shift:
                node_to_shift.left = node.left
            if node.right is not node_to_shift:
                node_to_shift.right = node.right

    def random(self):
        node = self.root

        # empty tree => None
        if not node:
            return

        while node:
            # a random float between 0.0 and 1.0
            n = random()
            total = 1 + node.left_cardinality + node.right_cardinality

            if n <= 1 / total:
                return node
            elif n <= (1 + node.left_cardinality) / total:
                node = node.left
            else:
                node = node.right
